### PR TITLE
fix: improve wire decorator types

### PR DESCRIPTION
**Pull request by sfdc-jkovitz**
## Details
Integrates some suggested type additions/modifications which should provide better type safety around the wire decorator

## Does this PR introduce breaking changes?

* ✅ `No, it does not introduce breaking changes.`

## GUS work item
W-8065114

### CODE REVIEW

File: packages/lwc/types.d.ts [hunk 1]
```diff
@@ -195,33 +195,34 @@ declare module 'lwc' {
...
     export function wire(getType: (config?: any) => any, config?: any): PropertyDecorator;
 
-    type WireConfigValue = Record<string, any>;
-    type ContextValue = Record<string, any>;
-
-    interface WireAdapter {
-        update(config: WireConfigValue, context?: ContextValue): void;
-        connect(): void;
-        disconnect(): void;
+    interface WireAdapterConstructor<Value, Config, Context> {
```
Comment (jodarove) [inline packages/lwc/types.d.ts:198]: i do like this generics use, i have a few questions and a request:

- should this types match the ones exported by https://github.com/salesforce/lwc/blob/master/packages/%40lwc/engine-core/src/framework/main.ts#L45-L52  ?
- do you think is worth to replace the existing types in `@lwc/engine-core` with these generic ones?

can you create some examples of code using this new types?

**Review by jodarove — CHANGES_REQUESTED** [review]
Does this fixes https://github.com/salesforce/lwc/issues/2026 ?

**Comment by jodarove** [comment]
@sfdc-jkovitz can you add to the description of the PR that this pr is fixing #2026?

File: packages/lwc/types.d.ts [hunk 2]
```diff
@@ -195,33 +195,34 @@ declare module 'lwc' {
      */
     export function wire(getType: (config?: any) => any, config?: any): PropertyDecorator;
```
Comment (jodarove) [inline packages/lwc/types.d.ts:196]: #2026 mentions an issue with the type of `@wire` decorator; im struggling to understand how the changes in this PR will fix that issue. can you explain more?

**Review by jodarove — CHANGES_REQUESTED** [review]

File: packages/lwc/types.d.ts (hunk 2 above)
Comment (sfdc-jkovitz) [reply packages/lwc/types.d.ts]: I've updated this, and the error has gone away, but let's discuss when you have a moment?

**Comment by jodarove** [comment]
closed in favor of https://github.com/salesforce/lwc/pull/2167

**Pull request closed without merging**